evm/chains/sharding/mainchain_handler/vmc_handler.py

```diff
@@ -0,0 +1,186 @@
+import logging
+
+from web3.contract import (
+    Contract,
+)
+
+from eth_utils import (
+    is_canonical_address,
+    to_checksum_address,
+    to_dict,
+)
+
+from evm.utils.hexadecimal import (
+    decode_hex,
+)
+
+from evm.chains.sharding.mainchain_handler.config import (
+    DEPOSIT_SIZE,
+    GASPRICE,
+    TX_GAS,
+)
+
+
+class VMC(Contract):
+
+    logger = logging.getLogger("evm.chain.sharding.mainchain_handler.VMC")
+
+    @to_dict
+    def mk_build_transaction_detail(self,
+                                    nonce,
+                                    gas,
+                                    chain_id=None,
+                                    value=None,
+                                    gas_price=None,
+                                    data=None):
+        if not (isinstance(nonce, int) and nonce >= 0):
+            raise ValueError('nonce should be provided as non-negative integer')
+        if not (isinstance(gas, int) and gas > 0):
+            raise ValueError('gas should be provided as positive integer')
+        yield 'nonce', nonce
+        yield 'gas', gas
+        yield 'chainId', chain_id
+        if value is not None:
+            yield 'value', value
+        if gas_price is not None:
+            yield 'gasPrice', gas_price
+        if data is not None:
+            yield 'data', data
+
+    def send_transaction(self,
+                         func_name,
+                         args,
+                         privkey,
```

Comment:
Seeing the private key passed around makes me think it would be cleaner if this was set as a property on the contract instance.

You can override the `__init__` as follows:

```python
class VMC(Contract):
    def __init__(self, *args, private_key, **kwargs):
        self.private_key = private_key
        super().__init__(*args, **kwargs)
```

Reply:
No problem. I will change it.